src/provider/shplonk.rs

```diff
@@ -284,23 +266,35 @@ where
     // P_i+1(r^2) == (1 - point_i) * P_i_even + point_i * P_i_odd -> should hold, according to Gemini transformation
     let mut point = point.to_vec();
     point.reverse();
+
     #[allow(clippy::disallowed_methods)]
-    for (index, ((eval_r, eval_minus_r), eval_r_squared)) in evals_at_r
-      .iter()
-      .zip_eq(evals_at_minus_r.iter())
-      // TODO: Ask Adrian if we need evals_at_r_squared[0] for some additional checks
-      .zip(evals_at_r_squared[1..].iter())
+    let verification_failed = pi.evals[0]
+      .par_iter()
+      .chain(&[*P_of_x])
+      .zip_eq(pi.evals[1].par_iter().chain(&[*P_of_x]))
+      .zip(pi.evals[2][1..].par_iter().chain(&[*P_of_x]))
       .enumerate()
-    {
-      let even = (*eval_r + eval_minus_r) * (E::Fr::from(2).invert().unwrap());
-      let odd = (*eval_r - eval_minus_r) * ((E::Fr::from(2) * r).invert().unwrap());
+      .any(|(index, ((eval_r, eval_minus_r), eval_r_squared))| {
+        let even = (*eval_r + eval_minus_r) * (E::Fr::from(2).invert().unwrap());
+        let odd = (*eval_r - eval_minus_r) * ((E::Fr::from(2) * r).invert().unwrap());
 
-      if *eval_r_squared != ((E::Fr::ONE - point[index]) * even) + (point[index] * odd) {
-        return Err(NovaError::ProofVerifyError);
-      }
+        *eval_r_squared != ((E::Fr::ONE - point[index]) * even) + (point[index] * odd)
+      });
+
+    if verification_failed {
+      return Err(NovaError::ProofVerifyError);
     }
 
-    let C_P: E::G1 = pi.comms.iter().map(|comm| comm.to_curve()).rlc(&q);
+    // TODO: revisit C_P computing later with progress in https://github.com/huitseeker/arecibo/tree/rayon-parscan
+    let q_powers = HyperKZG::<E, NE>::batch_challenge_powers(q, pi.comms.len());
+    #[allow(clippy::redundant_closure)]
+    let C_P: E::G1 = pi
+      .comms
+      .par_iter()
+      .zip_eq(q_powers.par_iter())
+      .fold(|| E::G1::identity(), |acc, (comm, q_i)| acc + *comm * q_i)
+      .sum::<E::G1>();
```

Comment:
Parallel iteration here may be slower than just single threaded.

Reply:
Nope. 

Benchmarks for multiple threads (current):
```
PCS-Verifying 10/arecibo::provider::shplonk::EvaluationEngine<halo2curves::bn256::engine::Bn256, are...                                                                            
                        time:   [2.0567 ms 2.0857 ms 2.1172 ms]
                        change: [-1.4539% -0.7515% +0.2534%] (p = 0.13 > 0.05)
                        No change in performance detected.
                        
PCS-Verifying 12/arecibo::provider::shplonk::EvaluationEngine<halo2curves::bn256::engine::Bn256, are...                                                                            
                        time:   [2.0434 ms 2.0459 ms 2.0486 ms]
                        change: [-48.813% -25.791% -3.5808%] (p = 0.20 > 0.05)
                        No change in performance detected.

PCS-Verifying 14/arecibo::provider::shplonk::EvaluationEngine<halo2curves::bn256::engine::Bn256, are...                                                                            
                        time:   [2.0393 ms 2.0443 ms 2.0497 ms]
                        change: [-2.5698% -2.3906% -2.1855%] (p = 0.00 < 0.05)
                        Performance has improved.

PCS-Verifying 16/arecibo::provider::shplonk::EvaluationEngine<halo2curves::bn256::engine::Bn256, are...                                                                            
                        time:   [2.0812 ms 2.1236 ms 2.2181 ms]
                        change: [+1.2374% +5.4436% +9.5572%] (p = 0.02 < 0.05)
                        Performance has regressed.

PCS-Verifying 18/arecibo::provider::shplonk::EvaluationEngine<halo2curves::bn256::engine::Bn256, are...                                                                            
                        time:   [2.1816 ms 2.1905 ms 2.2018 ms]
                        change: [-2.2863% -1.7509% -1.2855%] (p = 0.00 < 0.05)
                        Performance has improved.

PCS-Verifying 20/arecibo::provider::shplonk::EvaluationEngine<halo2curves::bn256::engine::Bn256, are...                                                                            
                        time:   [2.2137 ms 2.2172 ms 2.2250 ms]
                        change: [-1.7976% -1.4260% -1.0409%] (p = 0.00 < 0.05)
                        Performance has improved.

```

Using following single-threaded computation:
```
let C_P: E::G1 = comms
      .iter()
      .zip_eq(q_powers.iter())
      .fold(E::G1::identity(), |acc, (comm, q_i)| acc + *comm * q_i);
```

benches are following:
```
PCS-Verifying 10/arecibo::provider::shplonk::EvaluationEngine<halo2curves::bn256::engine::Bn256, are...                                                                            
                        time:   [3.1346 ms 3.1358 ms 3.1372 ms]
                        change: [+50.333% +51.841% +52.907%] (p = 0.00 < 0.05)
                        Performance has regressed.

PCS-Verifying 12/arecibo::provider::shplonk::EvaluationEngine<halo2curves::bn256::engine::Bn256, are...                                                                            
                        time:   [3.4385 ms 3.4397 ms 3.4407 ms]
                        change: [+67.825% +68.108% +68.408%] (p = 0.00 < 0.05)
                        Performance has regressed.

PCS-Verifying 14/arecibo::provider::shplonk::EvaluationEngine<halo2curves::bn256::engine::Bn256, are...                                                                            
                        time:   [3.7515 ms 3.7609 ms 3.7826 ms]
                        change: [+83.945% +85.587% +87.368%] (p = 0.00 < 0.05)
                        Performance has regressed.

PCS-Verifying 16/arecibo::provider::shplonk::EvaluationEngine<halo2curves::bn256::engine::Bn256, are...                                                                            
                        time:   [4.0919 ms 4.1829 ms 4.2782 ms]
                        change: [+78.136% +85.848% +93.416%] (p = 0.00 < 0.05)
                        Performance has regressed.

PCS-Verifying 18/arecibo::provider::shplonk::EvaluationEngine<halo2curves::bn256::engine::Bn256, are...                                                                            
                        time:   [4.4307 ms 4.4367 ms 4.4464 ms]
                        change: [+101.58% +102.62% +103.65%] (p = 0.00 < 0.05)
                        Performance has regressed.

PCS-Verifying 20/arecibo::provider::shplonk::EvaluationEngine<halo2curves::bn256::engine::Bn256, are...                                                                            
                        time:   [4.7740 ms 4.7775 ms 4.7831 ms]
                        change: [+113.80% +114.74% +115.62%] (p = 0.00 < 0.05)
                        Performance has regressed.
```

there is a linear grow of complexity for single threaded `C_P` computing with increasing number of variables